modules/mempressure.c

```diff
@@ -0,0 +1,754 @@
+ */
+static int
+mempressure_limit_repr(const mempressure_limit_t *self, char *data, size_t size)
+{
+    int res = snprintf(data, size, "used %d", self->mnl_used);
+    return res;
+}
+
+/** Parse limit object from /sys/fs/cgroup/memory/memory.usage_in_bytes format
+ */
+static bool
+mempressure_limit_parse(mempressure_limit_t *self, const char *data)
+{
+    char     *end = 0;
+    uint64_t  val = strtoull(data, &end, 10);
+    bool      res = end > data && *end == 0;
+
+    if( !res )
+        mce_log(LL_ERR, "parse error: '%s' is not a number", data);
+    else
+        self->mnl_used = mempressure_bytes_to_pages(val);
+
+    return res;
+}
+
+/** Check if limit object values are exceeded by given state data
+ */
+static bool
+mempressure_limit_exceeded(const mempressure_limit_t *self,
+                           const mempressure_limit_t *status)
+{
+    return (mempressure_limit_is_valid(self) &&
+            self->mnl_used <= status->mnl_used);
+}
+
+/* ========================================================================= *
+ * MEMPRESSURE_STATUS
+ * ========================================================================= */
+
+/** Configuration limits for normal/warning/critical levels */
+static mempressure_limit_t mempressure_limit[] =
+{
+    [MEMNOTIFY_LEVEL_NORMAL] = {
+        .mnl_used   = 0,
+    },
+    [MEMNOTIFY_LEVEL_WARNING] = {
+        // values come from config - disabled by default
+        .mnl_used   = 0,
+    },
+    [MEMNOTIFY_LEVEL_CRITICAL] = {
+        // values come from config - disabled by default
+        .mnl_used   = 0,
+    },
+};
+
+/** Cached status read from kernel device */
+static mempressure_limit_t mempressure_status =
+{
+    .mnl_used   = 0,
+};
+
+/** Cached memory use level */
+static memnotify_level_t mempressure_level = MEMNOTIFY_LEVEL_UNKNOWN;
+
+/** Check current memory status against triggering levels
+ */
+static memnotify_level_t
+mempressure_status_evaluate_level(void)
+{
+    memnotify_level_t res = MEMNOTIFY_LEVEL_UNKNOWN;
+
+    if( mempressure_limit_is_valid(&mempressure_status) ) {
+        res = MEMNOTIFY_LEVEL_NORMAL;
+
+        for( memnotify_level_t lev = MEMNOTIFY_LEVEL_NORMAL + 1;
+             lev < G_N_ELEMENTS(mempressure_limit); ++lev ) {
+            if( mempressure_limit_exceeded(mempressure_limit+lev, &mempressure_status) )
+                res = lev;
+        }
+    }
+
+    return res;
+}
+
+/** Re-evaluate memory use level and broadcast changes via datapipe
+ */
+static bool
+mempressure_status_update_level(void)
+{
+    memnotify_level_t prev = mempressure_level;
+    mempressure_level = mempressure_status_evaluate_level();
+
+    if( mempressure_level == prev )
+        goto EXIT;
+
+    mce_log(LL_WARN, "mempressure_level: %s -> %s",
+            memnotify_level_repr(prev),
+            memnotify_level_repr(mempressure_level));
+
+    datapipe_exec_full(&memnotify_level_pipe
+                       , GINT_TO_POINTER(mempressure_level));
```

Comment:
A little inconsistent formatting, usually "," seems to be at the end of previous line.

Reply:
fixed